src/coroutine/cancel.rs

```diff
@@ -0,0 +1,68 @@
+use crate::{ffi, Py, PyObject, Python};
+use futures_util::future::poll_fn;
+use futures_util::task::AtomicWaker;
+use std::ptr;
+use std::sync::atomic::{AtomicPtr, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+#[derive(Debug, Default)]
+struct Inner {
+    exception: AtomicPtr<ffi::PyObject>,
+    waker: AtomicWaker,
+}
+
+/// Helper used to wait and retrieve exception thrown in coroutine.
+#[derive(Debug, Default)]
+pub struct CoroutineCancel(Arc<Inner>);
+
+impl CoroutineCancel {
+    /// Create a new `CoroutineCancel`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Return an associated [`CancelHandle`].
+    pub fn handle(&self) -> CancelHandle {
+        CancelHandle(self.0.clone())
+    }
+
+    fn take_exception(&self) -> PyObject {
+        let ptr = self.0.exception.swap(ptr::null_mut(), Ordering::Relaxed);
+        Python::with_gil(|gil| unsafe { Py::from_owned_ptr(gil, ptr) })
+    }
+
+    /// Returns whether the associated coroutine has been cancelled.
+    pub fn is_cancelled(&self) -> bool {
+        !self.0.exception.load(Ordering::Relaxed).is_null()
+    }
+
+    /// Poll to retrieve the exception thrown in the associated coroutine.
+    pub fn poll_cancelled(&mut self, cx: &mut Context<'_>) -> Poll<PyObject> {
+        if self.is_cancelled() {
+            return Poll::Ready(self.take_exception());
+        }
+        self.0.waker.register(cx.waker());
+        if self.is_cancelled() {
+            return Poll::Ready(self.take_exception());
+        }
```

Comment:
I wonder, is all this complexity needed? Is a `oneshot` channel carrying the exception able to achieve the same effect?

Reply:
`oneshot` is one shot, but cancellation is not. Indeed, you can ignore it (see [last paragraph](https://docs.python.org/3/library/asyncio-task.html#task-cancellation)) and [uncancel](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.uncancel), so you may have several cancellation to handle.
That's the reason of this complexity.